src/supervisor.rs

```diff
@@ -26,59 +48,42 @@ pub struct StartError {
 #[derive(Debug)]
 pub struct TerminationError(HashMap<String, Arc<worker::TerminationError>>);
 
-pub struct Spec {
+pub struct SpecMeta {
     pub name: String,
     pub ev_notifier: EventNotifier,
     pub order: Order,
     pub strategy: Strategy,
+}
+
+pub struct Spec {
     pub children: Vec<worker::Spec>,
+
+    pub meta: SpecMeta,
 }
 
 pub struct Supervisor {
     pub runtime_name: String,
     pub children: Vec<Worker>,
 
-    pub name: String,
-    pub ev_notifier: EventNotifier,
-    pub order: Order,
-    pub strategy: Strategy,
-}
-
-fn ordered_children_on_start(
-    order: &Order,
-    children_specs: Vec<worker::Spec>,
-) -> Box<dyn Iterator<Item = worker::Spec>> {
-    match order {
-        Order::LeftToRight => Box::new(children_specs.into_iter()),
-        Order::RightToLeft => Box::new(children_specs.into_iter().rev()),
-    }
-}
-
-fn ordered_children_on_termination(
-    order: &Order,
-    children: Vec<Worker>,
-) -> Box<dyn Iterator<Item = Worker>> {
-    match order {
-        Order::LeftToRight => Box::new(children.into_iter().rev()),
-        Order::RightToLeft => Box::new(children.into_iter()),
-    }
+    pub meta: SpecMeta,
 }
 
 // terminate_prev_siblings gets called when a child in a supervision tree fails
 // to start, and all it's previous siblings need to get terminated.
-async fn terminate_prev_siblings<'a, 'b, 'c>(
+async fn terminate_prev_siblings<I: DoubleEndedIterator<Item = worker::Spec>>(
     ev_notifier: &mut EventNotifier,
+    order: Order,
     failed_child_spec: worker::Spec,
     failed_child_start_err: Arc<worker::StartError>,
-    pending_children_specs: Vec<worker::Spec>,
+    pending_children_specs: I,
     started_children: Vec<Worker>,
 ) -> (Vec<worker::Spec>, StartError) {
     let prev_sibling_termination_result =
         terminate_children(ev_notifier, started_children.into_iter().rev()).await;
 
     let (prev_specs, failing_sibling_termination_error) = match prev_sibling_termination_result {
-        Err((prev_specs, sibling_err)) => (prev_specs, Some(sibling_err)),
-        Ok(prev_specs) => (prev_specs, None),
+        (prev_specs, Err(sibling_err)) => (prev_specs, Some(sibling_err)),
+        (prev_specs, _) => (prev_specs, None),
     };
```

Comment:
Since all this does now is transform a `Result<(), TerminationError>` into an `Option<TerminationError>` you can get the same effect by just calling [`Result::err`](https://doc.rust-lang.org/std/result/enum.Result.html#method.err).

```rust
    let (prev_specs, failing_sibling_termination_result) =
        terminate_children(ev_notifier, started_children.into_iter().rev()).await;

    let failing_sibling_termination_error = failing_sibling_termination_result.err();
```